### DISCUSSION TITLE
What does PM mode actually mean?

### DISCUSSION THREAD

**Discussion by carlosthe19916** [guacsec/trustify#268] [General]
From the README file it seems what we called PM mode is:
```shell
cargo run --bin trustd
```

### Current status:

As of today, anyone who wants to start the so called "PM mode" needs:
- Clone the repository
  - git needs to be installed and `git clone` should be executed
- Have the appropriate Rust installed in the computer.
  - The user might use the office docs for it https://www.rust-lang.org/tools/install
- Have the appropriate NodeJS installed.
  - The user might use NVM https://github.com/nvm-sh/nvm?tab=readme-ov-file#install--update-script. But certainly there might be other ways the user might take.
- Finally the user will be able to execute `cargo run --bin trustd` and have the server running

### Risks
- Unexpectedly I discovered that our repository does not build on MacOS, neither on Windows. Here is a test I did https://github.com/trustification/trustify/pull/266 . So "PM" mode will work as long as PM only uses Linux.
- Today @ctron discovered that although he had the appropriate NodeJS version installed on Fedora 40, he couldn't build the server due to the UI not being able to be build. https://github.com/trustification/trustify/issues/265 and https://github.com/trustification/trustify-ui/issues/14 has the discussion records.
- As described in the previous section "Current status", although PM mode seems relatively easy for us (we work on the code every day), how easy or error prone would it be for another person?

### How to mitigate the risk
So how should we avoid the risk of having the issues while having multiple factors and environments a user can use?

- My instinct tell me: CONTAINER IMAGES. So the command `cargo run --bin trustd` runs in a container and not in the user's terminal.
- I know you can create binaries from Rust apps, so is the idea to create binaries for each OS so PM only downloads the binaries? If so, should we create binaries for each OS?

Those are just thoughts that came to my mind today. Feel free to disagree in any point stated.

**Comment by jcrossley3** [comment]
I agree it's more "dev mode" than "pm mode", but I still like it way more than what we had in the first project. Let's ask @mid998 what she'd prefer and go from there.

**Reply by carlosthe19916** [reply]
I agree. We currently have "dev mode", that is a very accurate naming. I actually like it and use it every day (I am a developer so I use "dev mode")... "pm mode" is yet to be defined.

**Comment by bobmcwhirter** [comment]
Also, when/once/if we set up a "release" job producing nightly binary releases, `pm mode` becomes a runtime thing only.  Sure, the build still requires build tools, but if someone downloads `trustify-osx.zip` and runs the contents, it should *just work*.

**Reply by bobmcwhirter** [reply]
`pm mode` really just just `good OOTB experience mode` for day-0 operations and test-drives.

**Reply by bobmcwhirter** [reply]
OOTB == "out of the box"

**Comment by gildub** [comment]
Context is everything, isn't ? 
My 2 cents guess : Physical Machine ?

**Comment by mid998** [comment]
pm mode should make you say “Damn, that was easy. Where’s my wine?"

Michelle DiPalma
Principal Product Manager
***@***.***

**Comment by helio-frota** [comment]
Can we close this?
